src/wp-includes/class-wp-block-bindings-registry.php

```diff
@@ -137,13 +163,48 @@ public function register( string $source_name, array $source_properties ) {
 			return false;
 		}
 
+		// Validate that the uses_context parameter is an array.
+		if ( isset( $source_properties['uses_context'] ) && ! is_array( $source_properties['uses_context'] ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'The "uses_context" parameter must be an array.' ),
+				'6.5.0'
+			);
+			return false;
+		}
+
+		if ( ! empty( array_diff( array_keys( $source_properties ), $this->allowed_source_properties ) ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				__( 'The $source_properties array contains invalid properties.' ),
+				'6.5.0'
+			);
+			return false;
+		}
+
 		$source = new WP_Block_Bindings_Source(
 			$source_name,
 			$source_properties
 		);
 
 		$this->sources[ $source_name ] = $source;
 
+		// Add `uses_context` defined by block bindings sources.
+		add_filter(
+			'get_block_type_uses_context',
+			function ( $uses_context, $block_type ) use ( $source ) {
```

Comment:
One more thing here. I believe we should use a class method and unregister the filter when unregistering the source. It would also help with documenting, as we could put all details from inline comments in the PHPDoc.

I'm not entirely sure how that would look in practice, but happy to explore together options.

Reply:
This makes sense to me. Although I don't know how it will work 🤔 I was thinking something like this:

```php
	/**
	 * Source that is being processed.
	 *
	 * @since 6.5.0
	 * @var WP_Block_Bindings_Source|false
	 */
	private $source;

	/**
	 * Callback passed to the `get_block_type_uses_context` filter to add the source's `uses_context` to the block.
	 *
	 * @since 6.5.0
	 *
	 * @param array         $uses_context Array of registered uses context for a block type.
	 * @param WP_Block_Type $block_type   The full block type object.
	 * @return array Modified array of uses_context with the values provided by the source.
	 */
	public function _add_source_uses_context( $uses_context, $block_type ) {
		$source = $this->source;
		if ( ! in_array( $block_type->name, $this->supported_blocks, true ) || empty( $source->uses_context ) ) {
			return $uses_context;
		}
		// Use array_values to reset the array keys.
		return array_values( array_unique( array_merge( $uses_context, $source->uses_context ) ) );
	}
```

But we probably need a different callback name per source.

Anyway, I'll add a task in [the tracking issue](https://github.com/WordPress/gutenberg/issues/53300) to handle it 🙂